PROTOCOL.md

```diff
@@ -427,63 +427,204 @@ Field    | Type | Description
 `errorMessage` | `string` | Human-readable message describing the error.
 `errorCode` | `string` | Machine-readable string describing the type of error.
 
-## Message Layer
+## Stream Layer
 
-The Message Layer contains three different types: `MessageID`, `MessageRef` and `StreamMessage`. This document describes message layer version `31`.
+This document describes Stream Layer version `32`.
 
-### StreamMessage
+Stream Layer messages are the actual content in streams. They are identified by a [`MessageID`](#messageid), are chained together via a reference to the previous message, [`MessageRef`](#messageref) to establish message order, they are cryptographically signed by their publisher. The most prominent Stream Layer message is the [`StreamMessage`](#streammessage), which is used to transport arbitrary application-level messages from publishers to subscribers. Additionally, there are a few other message types used for requesting and delivering encryption keys.
 
-Contains the data and metadata for a message produced/consumed on a stream. It is a payload at the Control Layer for the following message types: `PublishRequest`, `BroadcastMessage`, `UnicastMessage`. Where `msgId` uniquely identifies the `StreamMessage` and is the array representation of the `MessageID` defined [below](#messageid). `prevMsgRef` allows to identify the previous `StreamMessage` on the same stream and same partition published by the same producer. It is used to detect missing messages. It is the array representation of the `MessageRef` defined [below](#messageref).
+It's worth noting that the following Control Layer messages contain a Stream Layer message: `PublishRequest`, `BroadcastMessage`, and `UnicastMessage`.
 
-```
-[version, msgId, prevMsgRef, contentType, encryptionType, content, signatureType, signature]
-```
-Example:
-```
-[31, [...msgIdFields], [...msgRefFields], 27, 0, "contentData", 1, "0x29c057786Fa..."]
-```
+All Stream Layer messages have the following structure:
 
-Field    | Type | Description
--------- | ---- | --------
-`version` | `number` | Is currently 30.
-`msgId` | MessageID |Array representation of the `MessageID` to uniquely identify this message. 
-`prevMsgRef` | MessageRef | Array representation of the `MessageRef` of the previous message on a message chain (defined in the `msgId`). Used to detect missing messages.
-`contentType` | `number` | Determines how the content should be parsed according to the table below.
-`encryptionType` | `number` | Encryption type as defined by the table below.
-`content` | `string` | Content data of the message.
-`signatureType` | `number` | Signature type as defined by the table below.
-`signature` | `string` | Signature of the message, signed by the producer. Encoding depends on the signature type.
+ ```
+ [version, msgId, prevMsgRef, messageType, contentType, encryptionType, encryptionKeyId, content, signatureType, signature]
```

Comment:
```suggestion
 [version, msgId, prevMsgRef, messageType, contentType, encryptionType, groupKeyId, content, signatureType, signature]
```

IMO we **should** rename all the `groupKey` stuff to `encryptionKey` since  "groupKey" being related to encryption at all is not immediately obvious, unlike "encryptionKey", at least to me.

Reply:
(also needs same change in PR description)

Reply:
I don't want to snowball this at this time, so leaving the naming as is for now. "Group key" isn't by any means incorrect, it's even "more" correct (accurate) than "encryption key". Can be reconsidered in the future.

The important thing IMO here is that we make sure that the interface of the library is friendly enough to be used by users not familiar with cryptography. The spec doesn't need to be written in layman terms.

The only thing that's "wrong" now is that we're "abusing" the `groupKeyId` field to identify the RSA key used in case of RSA encryption. That is not a group key. But even so, I think I'll leave it as is for now.